src/mondefense.cpp

```diff
@@ -1,38 +1,47 @@
+#include "mondefense.h"
 #include "monster.h"
+#include "creature.h"
+#include "damage.h"
 #include "game.h"
 #include "rng.h"
 #include "line.h"
 #include "bodypart.h"
-#include "speech.h"
 #include "messages.h"
-#include "mondefense.h"
-#include <algorithm>
 
-
-void mdefense::zapback(monster *m, Creature *source, const projectile *proj)
+void mdefense::zapback(monster *const m, Creature *const source, projectile const* const proj)
 {
-    player *foe = dynamic_cast< player* >( source );
-    monster *othermon = dynamic_cast< monster* >( source );
-    if( rl_dist( m->pos(), source->pos() ) > 1 ) {
-        return; // Out of range
+    if ((proj) ||                              // Not a melee attack
+        (rng(0, 100) > m->def_chance) ||       // Attacker lucked out
+        (rl_dist(m->pos(), source->pos()) > 1) // Out of range
+    ) {
+        return;
     }
-    if( proj != nullptr || rng(0, 100) > m->def_chance ) {
-        return; // Not a melee attack or attacker lucked out
+
+    auto const is_eligible_player = [](player const *const foe) {
```

Comment:
Can't use a scope if you need to get a result out of it, or if you want to avoid uninitialized or double initialized variables.

```cpp
bool b; // or false or true
{
    // calculations...
    b = result;
}

// (1) oops, b isn't const despite semantically being so.
// (2) what if b weren't a bool but some more expensive to construct type?
// (3) what if b weren't a bool and had no sensible default state?
```

And, FWIW, lambdas don't imply indirection, especially locally defined lambdas: with any sort of optimization whatsoever turned on these evaporate into nothingness. They usually only imply a cost  if they're going to be copied around, or if they capture by value. I look at local lambdas (among their other uses of course) as a way to give a chunk of code a name, and to allow cleaner initialization (due to more flexible control flow).